agent/php_api.c

```diff
@@ -1578,3 +1578,73 @@ PHP_FUNCTION(newrelic_get_trace_metadata) {
     nr_free(span_id);
   }
 }
+
+/*
+ * Purpose      (New Relic API) (Error Fingerprinting)
+ *              Allows a user to register a PHP callback to be invoked by the
+ *              Agent when the application encounters an error for the purpose
+ *              of returning a string for the 'error.group.name' Agent
+ *              Attribute.
+ *
+ */
+#ifdef TAGS
+void zif_newrelic_set_error_group_callback(void); /* ctags landing pad only */
+void newrelic_set_error_group_callback(void);     /* ctags landing pad only */
+#endif
+PHP_FUNCTION(newrelic_set_error_group_callback) {
+  zend_fcall_info fci;
+  zend_fcall_info_cache fcc;
+
+  NR_UNUSED_RETURN_VALUE_PTR;
+  NR_UNUSED_THIS_PTR;
+  NR_UNUSED_RETURN_VALUE_USED;
+
+  nr_php_api_add_supportability_metric("set_error_group_callback" TSRMLS_CC);
+
+  // Verify that only one argument has been provided to the API (the callback)
+  if (1 != ZEND_NUM_ARGS()) {
+    nrl_warning(NRL_API,
+                "newrelic_set_error_group_callback failure: invalid number of "
+                "parameters");
+    RETURN_FALSE;
+  }
+
+  // Verify that the argument passed to the API is a function
+  // Populate function data
+  if (FAILURE == zend_parse_parameters(ZEND_NUM_ARGS(), "f", &fci, &fcc)) {
+    nrl_warning(
+        NRL_API,
+        "newrelic_set_error_group_callback failure: invalid argument passed");
+    RETURN_FALSE;
+  }
+
+  // Verify the user callback accepts 2 arguments
+  if (2 != fcc.function_handler->common.num_args) {
+    nrl_warning(NRL_API,
+                "newrelic_set_error_group_callback failure: invalid number of "
+                "callback parameters: %d",
+                fcc.function_handler->common.num_args);
+    RETURN_FALSE;
+  }
+
+  // Log debug message if the user is overwriting an existing callback
+  if (NULL != NRPRG(error_group_user_callback)) {
+    nrl_debug(
+        NRL_API,
+        "newrelic_set_error_group_callback: overwriting previous callback");
```

Comment:
Should the `error_group_user_callback` be overwritten or left as it was?

Reply:
According to the Agent Spec, any subsequent calls to the API to register a callback will clobber any previously registered callbacks.

Reply:
Then I think this code needs to be updated as it currently doesn't overwrite `error_group_user_callback`.

Reply:
Also would be good to have a test for this case, i.e. that the callback gets overwritten.

Reply:
Ah! I think I know what is happening - this code only allocates memory if it was not allocated. Then the comment is misleading a bit. And I found the test for 'clobbering'.

Reply:
Yes, this is just a message for debugging purposes in case someone has multiple callbacks and can't understand why it isn't behaving how they might expect. The memory is already allocated, so we can just overwrite  `fci` and `fcc` directly without having to reallocate. I have included an integration test that tests this functionality.